content/lectures/lec02.md

```diff
@@ -6,6 +6,80 @@ assignments: ["hw0"]
 draft: false
 ---
 
+# Lecture Two: Docker
+
+## Portability
+
+Recall all of the issues that we've run into trying transfer code that we developed on our local machine to a remote host in the cloud. We want to be able to take our application code, and seamlessly run it on another machine. Along with Python dependencies, we needed to make sure that system resources, including Python and Poetry themselves, were also installed as expected. How can we avoid all of the dependency and versioning issues we ran into when we manually deployed to EC2?
+
+We were able to use Poetry and get all of the *reproducibility* benefits when we deployed to EC2, but just because a build is reproducible does *not* mean that it's **portable**.
+
+The intuition here is that we essentially want to _send a copy of our local computer_ to whoever is going to be running our code.
+
+## Containers
+
+Containers allow us to package minimal operating systems as "images" which includes everything our program needs to run. Once we've created the image, we can send it to anyone, and they will be able to build it on their machine to run our code. If you've used Virtual Machines in previous classes, you can think of container images as analogous to VM images, but with much less overhead.
+
+We can also build the image locally and send the image after it has already been built. This way the receiver can run the container without having to worry about any of the build process.
+
+Ideally, the build process itself should also be reproducible. 
+
+## Docker
+
+Docker is the industry-standard implementation for containers. While there are other container implementations, Docker has become something like the "Kindle" or "Kleenex" of container implementations, but there [are](https://coreos.com/rkt/) [other](https://katacontainers.io) [implementations](https://firecracker-microvm.github.io) which have emerged recently and look to be giving Docker a run for its money. 
+
+The Open Container Image (OCI) format is a general specification for how to build container images originally based on the Docker image format. But other tools built against the OCI specification build compatible containers or run compatible containers, without running any code built by the Docker team. Implementations like [podman](https://containerd.io) and [containerd](https://containerd.io) are compatible with the OCI specification while improving on some ambiguities in the original Docker implementation. Kubernetes has actually [moved away from the Docker implementation](https://kubernetes.io/blog/2020/12/02/dont-panic-kubernetes-and-docker/) recently. We'll use "OCI container", "container", and "Docker container" interchangeably throughout these notes.
```

Comment:
OCI is Open Container Initiative, not Image. Correct wording would probably be "the OCI Image format"

Reply:
I'd also condense the 2nd sentence to have "build or run compatible containers"

Reply:
The Kubernetes Docker thing may cause more confusion than it's worth at the moment